From 173cbab44b48ea708cbe41fe8f4091c227f9a019 Mon Sep 17 00:00:00 2001
From: Ethan Chen <intervigil@gmail.com>
Date: Tue, 24 Jul 2018 20:04:50 -0700
Subject: [PATCH 3/3] Updater: Use SharedPreference listener to get perf mode
 setting

* Don't query shared preferences on each call to install.

Change-Id: I022d76ba0659d9ad3fac48d84fa4172476cb506b
---
 .../updater/controller/ABUpdateInstaller.java | 23 ++++++++++++++++---
 1 file changed, 20 insertions(+), 3 deletions(-)

diff --git a/src/org/lineageos/updater/controller/ABUpdateInstaller.java b/src/org/lineageos/updater/controller/ABUpdateInstaller.java
index b0be201..aeb77e4 100644
--- a/src/org/lineageos/updater/controller/ABUpdateInstaller.java
+++ b/src/org/lineageos/updater/controller/ABUpdateInstaller.java
@@ -53,6 +53,7 @@ class ABUpdateInstaller {
 
     private UpdateEngine mUpdateEngine;
     private boolean mBound;
+    private boolean mPerfModeEnabled;
 
     private boolean mFinalizing;
     private int mProgress;
@@ -119,6 +120,16 @@ class ABUpdateInstaller {
         }
     };
 
+    private final SharedPreferences.OnSharedPreferenceChangeListener mPrefListener =
+        new SharedPreferences.OnSharedPreferenceChangeListener() {
+        @Override
+        public void onSharedPreferenceChanged(SharedPreferences sharedPrefs, String key) {
+            if (Constants.PREF_AB_PERF_MODE.equals(key)) {
+                mPerfModeEnabled = isPerfModeEnabled(mContext);
+            }
+        }
+    };
+
     static synchronized boolean isInstallingUpdate(Context context) {
         SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
         return pref.getString(ABUpdateInstaller.PREF_INSTALLING_AB_ID, null) != null ||
@@ -147,10 +158,18 @@ class ABUpdateInstaller {
         return TextUtils.equals(waitingId, downloadId);
     }
 
+    static synchronized boolean isPerfModeEnabled(Context context) {
+        return PreferenceManager.getDefaultSharedPreferences(context)
+            .getBoolean(Constants.PREF_AB_PERF_MODE, false);
+    }
+
     private ABUpdateInstaller(Context context, UpdaterController updaterController) {
         mUpdaterController = updaterController;
         mContext = context.getApplicationContext();
         mUpdateEngine = new UpdateEngine();
+        mPerfModeEnabled = isPerfModeEnabled(mContext);
+        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(mContext);
+        prefs.registerOnSharedPreferenceChangeListener(mPrefListener);
     }
 
     static synchronized ABUpdateInstaller getInstance(Context context,
@@ -214,9 +233,7 @@ class ABUpdateInstaller {
             }
         }
 
-        boolean enableABPerfMode = PreferenceManager.getDefaultSharedPreferences(mContext)
-                .getBoolean(Constants.PREF_AB_PERF_MODE, false);
-        mUpdateEngine.setPerformanceMode(enableABPerfMode);
+        setPerformanceMode(mPerfModeEnabled);
 
         String zipFileUri = "file://" + file.getAbsolutePath();
         mUpdateEngine.applyPayload(zipFileUri, offset, 0, headerKeyValuePairs);
-- 
2.17.1

